daemon/c_smartcard.c

```diff
@@ -1075,13 +1083,22 @@ c_smartcard_new(compartment_t *compartment)
 		}
 	}
 
+	if (smartcard->token_type == CONTAINER_TOKEN_TYPE_PKCS11) {
+		if (NULL == container_get_pkcs11_module(smartcard->container)) {
+			ERROR("PKCS#11 module missing in container config. Abort creation of container");
+			mem_free0(smartcard);
+		}
+		smartcard->pkcs11_module =
+			mem_strdup(container_get_pkcs11_module(smartcard->container));
+	}
+
 	smartcard->success_cb = NULL;
 	smartcard->err_cb = NULL;
 	smartcard->err_cbdata = NULL;
 
 	smartcard->sock = sock_unix_create_and_connect(SOCK_SEQPACKET, scd_sock_path);
 	if (smartcard->sock < 0) {
-		mem_free0(smartcard);
+		mem_free0(smartcard); //? potential mem_leak for usbtoken
```

Comment:
yes smartcard->token_serial may be allocated here. Thus should be freed, same counts for smartcard->pkcs11_module. May instead of return NULL in error cases. We introduce a jump mark err:
and do cleanup there

Reply:
Done